Allow custom root directory for config (#8237)
astropy/astropy: docs/config.rst

parameters used in astropy or affiliated packages without delving into the
source code to make those changes.

Getting Started
===============

The Astropy configuration options are most easily set by modifying the

> configuration file. If you want to modify the configuration file from the default you can save the default config to a file::

Reviewer: ```suggestion
configuration file. If you want to modify the configuration file from
the default you can save the default config to a file::
```

Each time you upgrade to a new major version of astropy, the
configuration parameters may have changed.

If you never edited your configuration file, there is nothing for you to do. It
will automatically be replaced with a configuration file template for the newly

> installed version of astropy when running ``astropy.write_default_config``

Reviewer: ```suggestion
installed version of astropy when running ``astropy.write_default_config()``.
```

Each time you upgrade to a new major version of ``astropy``, the
configuration parameters may have changed.

> If you never edited your configuration file, there is nothing for you to do. It

Reviewer: "there is nothing for you to do" is not entirely accurate anymore? In fact, if users have been relying on `astropy.cfg`, nowt they have to explicitly run `astropy.write_default_config()` to get an updated `astropy.cfg`. Otherwise, the old one stays and might mess things up. Did I understand this new behavior correctly?

> will automatically be replaced with a configuration file template for the newly

Reviewer: "automatically" is also misleading now, as generating a new `astropy.cfg` requires manual intervention.

        return (conf.some_setting if val is None else val) + 2

.. _config-affiliated:

Affiliated Packages
===================

> If you maintain an affiliated package you can customise the behaviour of the

Reviewer: US English :smile: 
```suggestion
If you maintain an affiliated package, you can customize the behavior of the
```

> astropy configuration machinery for your package. Prior to version 4.0 of

Reviewer: ```suggestion
Astropy configuration machinery for your package. Prior to ``astropy`` 4.0,
```

> Astropy the bahviour of both ``astropy`` and the package template was to write a

Reviewer: ```suggestion
the behavior of both ``astropy`` and the package template was to write a
```

configuration file to ``~/.astropy/yourpackage.cfg`` if a template configuration

> file was provided by your package. If you wish to maintain this behaviour you do

Reviewer: ```suggestion
file was provided by your package. If you wish to maintain this behavior, you do
```

> not have to change anything.

Reviewer: I find that this paragraph directly contradicts the next one. Here, it says I don't have to change anything to retain the classic behavior, but the next paragraph says I need a wrapper to mimic what `astropy` is doing, which is also the classic behavior.

In addition, the first paragraph is also misleading, because when affiliated packages uses config system from `astropy>=4`, it does not automatically write on import anymore no matter how you configure it, you always have to run the wrapper below to generate a new file, right?

> If you want to change the behaviour of your package to mimic that of ``astropy``

Reviewer: ```suggestion
If you want to change the behavior of your package to mimic that of ``astropy``,
```

you can define a wrapper for `astropy.config.write_default_config` in your

> ``__init__.py`` file i.e.::

Reviewer: ```suggestion
``__init__.py`` file, i.e.::
```

  def write_default_config():
      """
      Writes out the template configuration file for this version of packagename.

      This function will save a template config file for manual editing, if a
      config file already exists, this will write a config file appended with the
      version number, to facilitate comparison of changes.

      Returns
      -------
      filepath : `pathlib.Path` or `None`
          The full path of the file written or `None` if no file was written.
      """
      from . import config as _config
      return _config.write_default_config('packagename')

This function provides a way for your users to write out a version of your
configuration template, in the same way that they are written on import before
this change.

Customizing the Configuration Directory
---------------------------------------

.. warning::

   If you used the :ref:`package template <packagetemplate:package-template>` to

>    create your affiliated package this might cause you to have to be more

Reviewer: This has to be reworded with clarification. What exactly the affected package has to be careful about? What signs of trouble should they be looking out for, etc?

   careful when pulling in updates from the package template until the template
   is updated to reflect these changes.

> Starting with the 4.0 release of astropy it is possible to override the name of

Reviewer: ```suggestion
Starting with the 4.0 release of ``astropy``, it is possible to override the name of
```

the directory where the configuration files are stored. Before 4.0 and by

> default this is the ``.astropy`` directory (often ``~/.astropy``). To change

Reviewer: ```suggestion
default, this was the ``.astropy`` directory (often ``~/.astropy``). To change
```

> this you can set the ``rootname=`` keyword in various places. To do this you

Reviewer: ```suggestion
this, you can set the ``rootname=`` keyword in various places.
```

> need to make two changes, firstly, you need to add a ``rootname='packagename'``

Reviewer: ```suggestion
Firstly, you need to add a ``rootname='packagename'``
```

argument to the ``_config.write_default_config`` call in the previous example.
Secondly you need to override the `astropy.config.ConfigItem` class. The best

> way to do that is to define a ``config`` submodule for your package i.e.

Reviewer: ```suggestion
way to do that is to define a ``config`` submodule for your package, i.e.,
```

``packagename/config/__init__.py`` would look like this::

  
  from astropy.config import ConfigNamespace, ConfigItem as _AstropyConfigItem

  __all__ = ['ConfigItem', 'ConfigNamespace']

  class ConfigItem(_AstropyConfigItem):
      rootname = 'packagename'

You would then import ``ConfigNamespace`` and ``ConfigItem`` from

> ``packagename.config``.

Reviewer: Also, maybe add a note that it is the affiliated package maintainer's responsibility to communicate this customization and that it means for the package's backward compatibility to their users. (So their users won't come here to complain because we can't do anything about it anyway.)
